README.md
# Just Good Cookies (JGC)

![JGC Styles](https://raw.githubusercontent.com/francescomugnai/just-good-cookies/main/images/jgc.jpg "JGC")

> **Please note**: JGC is in its **pre-release stage** It may be buggy and may not work as expected. Some features may be added, **others may be completely removed or delayed**.
> 🚧 It is currently being actively tested. Use at your own risk! 🚧

<br/>

![npm](https://img.shields.io/badge/Just%20Good%20Cookies-0.9.9-%23abe1be) ![jsDelivr hits (GitHub)](https://img.shields.io/jsdelivr/gh/hm/francescomugnai/just-good-cookies?color=b8eecc) ![GitHub top language](https://img.shields.io/github/languages/top/francescomugnai/just-good-cookies?color=%23bbf7d0) ![npm bundle size](https://img.shields.io/bundlephobia/min/just-good-cookies?color=%23bbf7d0) ![NPM](https://img.shields.io/npm/l/just-good-cookies?color=%23bbf7d0) ![GitHub Workflow Status](https://img.shields.io/github/workflow/status/francescomugnai/just-good-cookies/CodeQL?color=%23bbf7d0) ![Snyk Vulnerabilities for GitHub Repo](https://img.shields.io/snyk/vulnerabilities/github/francescomugnai/just-good-cookies?color=%23bbf7d0)

Just Good Cookies (JGC) is a simple (but delicious) cookie consent solution made in vanilla JavaScript.
It is designed to be fast, painless and easy to use.
JGC supports both automatic and manual cookie blocking modes, and also generates a fully automatic settings panel.

This library provides 9 different banner styles that **ONLY** work in conjunction with TailwindCSS ❤️ (which is not included in the package, so you can decide for yourself how to integrate it, see [below](#tailwind)).

## Key features

- 🍾 **Free and Open Source**: No pro plans, no subscriptions, no limits, no hidden fees, no surprises. I made it for myself, but I am happy to share it with everyone.
- 👌 **Standalone**: No external dependencies, just pure JS.
- 💅 **9 prebuilt styles ready to work with TailwindCSS**: You need to install Tailwind in order to use JGC.
  And it works with custom prefixes too, so if you prefer to use your own css (or another framework 😢) you can do that (and leave Tailwind only for the banner).
- ✏️ **Customization**: Change colors, positions, texts, styles, languages, buttons, behaviours, icons and much more!
- 🤙🏻 **Manual block**: JGC allows you to block cookies manually by defining data attributes.
- 🤖 **Auto mode**: This is a beta feature that allows you to block scripts and iFrames without having to mark them first. It definitely needs improvement, but it works great!
- 🎛 **Preference panel**: JGC generates a preference panel without you having to write a single line of code. You can even customize the [header and footer with plain html](<(#customizepanel)>). Give users the ability to revoke their consent!
- 🍪 **Cookie categories**: Defines different cookie categories with opt-in/out toggle (JGC will group them automatically in the preferences panel). You can also define some "necessary" cookies that users cannot turn off. Get granular cookie consent for categories!
- 🇪🇺 **GDPR compliant**: It's really easy to be compliant using JGC (it's up to you to comply with your state's regulations).
- 🇮🇹 **Multilingual**: Contributions welcome! Please, per favore, bitte, por favor, s'il vous plaît. I need your help 🙏
- 👨‍💻 **Developer friendly**: You can run the callbacks "onAccept" and "onReject" to store the proof of the consent (the backend is up to you, but it's pretty easy...maybe I can post an example in another repo if you want)
- 🌙 **Dark mode**: Supports dark mode according to Tailwind rules
- 🖼 **Someone said "placeholder"?**: This is something cool. You can define placeholders to hide the scripts until the user accepts the cookies. You can set a generic text, a default image or a custom one for each iFrame you block.

## Donations

JFC is a free open-source project that I work on in my spare time (but hey, I put a lot of work into it). <br/>
If you want to support the project, you can buy me a coffee or a beer (or two 🍻, I love beer) ❤️!
Thank you!

<a href="https://www.buymeacoffee.com/mugnai" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>

<hr/>

# Table of Contents

### Intro and examples

1. [Installation](#installation)
2. [How it works](#how)
3. [How to embed TailwindCSS](#tailwind)
4. [Full example](#full)
5. [Main Configuration](#main)
6. [Styles](#styles)

### Customization

7.  [Text](#customtext)
8.  [Banner](#banner)
9.  [Placeholder](#placeholder)
10. [Cookies](#cookies)
11. [Activate](#activate)
12. [Panel](#panel)

### Info

- [Disclaimer](#disclaimerArea)
- [ToDo](#toDo)

### WIKI - HOWTOs

- 🇮🇹🇺🇸🇪🇸 [How to translate the custom strings and make JGC fully multilingual](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-translate-the-custom-strings-and-make-JGC-fully-multilingual)
- 🇮🇹🇺🇸🇪🇸 [How to submit a new translation](<https://github.com/francescomugnai/just-good-cookies/wiki/How-to-submit-a-new-translation-(or-to-update-an-old-one)>)
- 🧑🏻‍💻 [How to customize the panel header and footer](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-customize-the-panel-header-and-footer)
- 🎛 [How to fire the preference panel from a button](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-fire-the-preference-panel)
- 👩🏼‍💻 [How to replace the banner description with a custom div](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-replace-the-banner-description-with-a-custom-div)
- 🖖🏻+🤖 [How to mix and merge the automatic mode and the manual one](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-mix-and-merge-the-automatic-mode-and-the-manual-one)
- 🖋 [How to use custom fonts](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-use-custom-fonts)
- 🖥 [How to avoid flickering](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-avoid-flickering)
- 🧙🏻 [How to install JGC on a website (eg. WordPress) that uses a different CSS framework](<https://github.com/francescomugnai/just-good-cookies/wiki/How-to-install-JGC-on-a-website-(eg.-WordPress)-that-uses-a-different-CSS-framework>)
- ❗️ [Browser support and how to deal with out-of-date browsers](https://github.com/francescomugnai/just-good-cookies/wiki/Browser-support-and-how-to-deal-with-out-of-date-browsers)

<hr>

<span name="installation"></span>

## Installation

There are 2 ways to use JGC in your projects:

- with the CDN
- as module

#### Using the CDN:

1. Copy this script and paste it on your site (ideally into the head tag)

```js
<script src="https://cdn.jsdelivr.net/gh/francescomugnai/just-good-cookies@0.9.9/dist/justgoodcookies.min.js"></script>
```

2. Install Tailwind (if you haven't already, see [below](#tailwind))

3. Initialize JGC

```js
<script>
justgoodcookies.init({
  locale: "it", // ISO 639-1 language code
  layout: "style1",
  privacyLink: "https://www.mysite.com/privacy-policy/",
  cookies: {
    necessary: { // this is mandatory
        title: "Necessary cookies",
        description: "Some text to describe them",
    },
  },
});
</script>
```

4. Tag the cookies you want to block (see ["How it works"](#how))

And...you are done 🚀

#### Using NPM

1. Install the package

```bash
npm install just-good-cookies
```

2. Import it into your bundle

```bash
import justgoodcookies from "just-good-cookies";
```

3. Install Tailwind (if you haven't already, see [below](#tailwind))

4. Initialize JGC (see the previous example) 🚀

<hr>

<span name="tailwind"></span>

## How to embed TailwindCSS

The easiest way to get started with JGC is to use Tailwind's default configuration (their documentation is amazing). You do not need to change anything.
Just download JGC (or install it using NPM) and write the path in the "content" object of your tailwind.config.js file.

However, if you want to use Tailwind with a **custom prefix** you must embed the new CDN: `<script src="https://cdn.tailwindcss.com"></script>`. <br/>
Then add the [`tailwindPrefix`](#tailwindPrefix) parameter in JGC.

Here's a basic example. <br/>

**Tailwind Config**:

```js
<script>
tailwind.config = {
  prefix: "jgc-",
};
</script>
```

**JGC**:

```js
justgoodcookies.init({
  tailwindPrefix: "jgc-",
  locale: "it", // ISO 639-1 language code
  layout: "style1",
  privacyLink: "https://www.mysite.com/privacy-policy/",
  cookies: {
    necessary: {
      title: "Necessary cookies",
      description: "Some text to describe them",
    },
  },
});
```

> JGC does not work with custom prefixes generated by JIT via Webpack, Rollup, Screw, and Parcel. For [this](https://tailwindcss.com/docs/content-configuration#dynamic-class-names) reason.

<span name="how"></span>

## How it works

Two options:

1. You can manually _tag_ the cookies you want to block.
2. You can use an automatic function to identify and block them.

#### Manual mode 🤙🏻

Let's see how to block a script or an iframe. It's pretty simple. <br/>
Get one, for example the Spotify widget:

```html
<iframe
  src="https://open.spotify.com/embed/album/1234567890"
  width="300"
  height="380"
  frameborder="0"
  allowtransparency="true"
  allow="encrypted-media"
></iframe>
```

Change it to:

```html
<iframe
  data-jgc-tag="thirdparties"
  data-jgc-src="https://open.spotify.com/embed/album/1234567890"
  data-jgc-service="Spotify widget"
  width="300"
  height="380"
  frameborder="0"
  allowtransparency="true"
  allow="encrypted-media"
></iframe>
```

Steps:

- replace `src` with `data-jgc-src`
- add `data-jgc-tag` to describe the type of cookie.
- use `data-jgc-service` to give the cookie a "name" (this value is returned in the settings panel).

And you're done. 🎉<br/>

#### Auto Mode 🤖

1. Initialize JGC and add two new properties: `autoMode` (boolean) and `autoCategories` (object)

```js
justgoodcookies.init({
  autoMode: true,
  autoCategories: {
    youtube: ['Youtube', 'functionalities'], // example
    spotify: ['Spotify', 'functionalities'], // example
    google: ['Google', 'analytics'], // example
    facebook: ['Facebook', 'marketing'], // example
  },
  locale: "en",
  layout: "style1",
  privacyLink: "https://www.mysite.com/privacy-policy/",
  cookies: {
    analytics: {
        title: "Analtycs or another title",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer pulvinar gravida urna quis vehicula. Vestibulum sed libero ultricies, facilisis nunc ut, accumsan magna.",
    },
    functionalities: {
        title: "Functionalities or another title",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer pulvinar gravida urna quis vehicula. Vestibulum sed libero ultricies, facilisis nunc ut, accumsan magna.",
    },
    necessary: {
        title: "Necessary cookies",
        description: "Lorem ipsum dolor sit amet, consectet",
    },
  },
)}
```

`autoMode` is self-explanatory. <br/>
`autoCategories` describes the scripts you want to block:

- The key is the **domain name of the script** (including the subdomain, if any)
- In the array you must specify **the name of the service** (which can be any string) and the **cookie type** (which must match the key specified in the `cookies` object).

Okay, enough talking — let's see a few examples.
Suppose you have a script like the following:

```html
<iframe
  src="https://www.youtube.com/embed/1234567890"
  width="560"
  height="315"
  title="YouTube video player"
  frameborder="0"
  allow="accelerometer; autoplay; clipboard-write;"
  allowfullscreen
>
</iframe>
```

To block it, you need to set `autoMode` to **true** and populate `autoCategories` in this way:

```js
autoMode: true,
autoCategories: {
    "www.youtube.com": ['Youtube Videos', 'functionalities'],
},
cookies: {
  functionalities: {
      title: "Functionalities or another title",
      description:
          "Lorem ipsum dolor sit amet, consectetur adipiscing elit etc...",
  },
  //...and so on
},
```

The key (_"www.youtube.com"_) is the domain name of the URL you are blocking.<br/>
In the array you have the _name_ of the service (_which is up to you_) and the _type_ (in this example: _functionalities_).
This type must be exactly the same key as that of the `cookie` object. <br/>
See how it works?

Ok, another example: a FB widget.

```html
<iframe
  src="https://www.facebook.com/plugins/...."
  scrolling="yes"
  style="border:none; overflow:hidden; width:300px; height:500px; background: white;  "
  allowtransparency="true"
  frameborder="0"
  data-src="https://www.facebook.com/plugins/..."
></iframe>
```

This is what you have to do:

```js
  autoCategories: {
      "www.facebook.com": ['Facebook Widget', 'marketing'],
  },
  cookies: {
    marketing: {
        title: "Marketing cookies",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer pulvinar gravida urna quis vehicula. Vestibulum sed libero ultricies, facilisis nunc ut, accumsan magna.",
    },
    //...and so on
  },
```

I hope that makes sense. I'll post more examples soon.

#### Limitations of the automatic mode

- It is possible that a script drops a cookie after a few milliseconds (which then disappears when JGC removes the iFrame). I still need to verify this point. Most of the scripts I have tested are blocked without problems.
- The placeholders only work with iFrames (because they have a specific width and height). Of course not with scripts.
- An automatic refresh is required to activate those scripts.

#### Placeholders

JGC allows you to define a placeholder (text or img) to overlay the blocked iFrames.<br/>
If you want to use a text placeholder add the attribute `data-jgc-placeholder-text`.

```html
<iframe
  data-jgc-placeholder-text="This is a placeholder"
  data-jgc-service="Spotify widget"
  data-jgc-tag="thirdparties"
  data-jgc-src="https://open.spotify.com/embed/album/1234567890"
  width="300"
  height="380"
  frameborder="0"
  allowtransparency="true"
  allow="encrypted-media"
></iframe>
```

Sometimes you may want to use an image as a placeholder.
In that case use the `data-jgc-placeholder-img` attribute.

```html
<iframe
  data-jgc-placeholder-img="https://images.unsplash.com/photo-1567095761054-7a02e69e5c43?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=987&q=80"
  data-jgc-service="Spotify widget"
  data-jgc-tag="thirdparties"
  data-jgc-src="https://open.spotify.com/embed/album/1234567890"
  width="300"
  height="380"
  frameborder="0"
  allowtransparency="true"
  allow="encrypted-media"
></iframe>
```

Now, suppose you are using autoMode (or manual mode) and want to load a generic placeholder over all blocked iFrames.
It's simple, add this object to your script:

```js
placeholder: {
  text: "This is a placeholder <br/> Click here to change preferences.",
  classes: 'bg-green-200 text-white bg-cover text-center', // use some classes to center or stylize the placeholder
}
```

or this one:

```js
placeholder: {
    classes: 'bg-cover',
    image: 'https://images.unsplash.com/photo-1607827448387-a67db1383b59?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2670&q=80'
}
```

How cool is that, huh?

<hr/>

#### BETA: The special attribute `data-jgc-remove`

In other situations, you will have to deal with scripts that are **almost impossible** to block (they release cookies even if you change the "src" attribute). <br/>
In these special cases, you can either use autoMode (hiding the scripts [with the css](#flickering)) or manual mode adding the `data-jgc-remove` attribute on the parent div.

A few notes about the `data-jgc-remove` implementation:

- The page will be refreshed once the user has accepted the cookies (because we remove the scripts from DOM and you can not reactivate them without refreshing).
- You can not use placeholders for these elements
- The two data attributes `data-jgc-service` and `data-jgc-tag` must be placed on the parent div (keeping the original script with the default attributes)

Example:

```html
<div
  data-jgc-remove
  data-jgc-service="Not so impossible to block after all"
  data-jgc-tag="marketing"
>
  <script
    type="text/javascript"
    src="https://impossiblesitetoblock.com/jsform/1234"
  ></script>
</div>
```

It's pretty simple 🥳

<span name="full"></span>

## Full example

Here is an example of **all** properties that can be set during initialization:

```js
<script>
  justgoodcookies.init({
    locale: "it",
    layout: "style1",
    privacyLink: "https://www.mysite.com/privacy-policy/",
    cookieDuration: 182, // in days, default cookie duration 360 days
    tailwindPrefix: "jgc-",
    dark: true,
    banner: {
      animation: true,
      backgroundColor:'bg-color',
      backgroundDark: true,
      backgroundImage: '/path/to/image.jpg',
      closeButton: false,
      closeButtonAccept: true,
      disableReject: false,
      icon: "/assets/cookies_icon.svg",
      iconDark: "/assets/cookies_icon_dark.svg",
      innerBackgroundImage: '/assets/cookies-style6.svg',
      logo: "/assets/logo.svg",
      logoClasses: 'w-2/4',
      maxWidth: 'xl',
      onAccept(){},
      onReject(){},
      position: "bottom",
      shortText: false,
      title: "Custom title",
    },
    cookies: {
      necessary: { // MANDATORY
        title: "Necessary cookies",
        description: "Lorem ipsum dolor sit amet,consectet",
      },
      // analytcs here is an example, just "necessary" is mandatory
      analytics: {
        title: "Analtycs or another title",
        description:
          "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer pulvinar gravida urna quis vehicula. Vestibulum sed libero ultricies, facilisis nunc ut, accumsan magna.",
      },
    },
    placeholder: {
      text: "This is a placeholder <br/> Click here to change preferences.",
      classes: 'bg-green-200 text-white bg-cover text-center',
      image: 'https://images.unsplash.com/photo-1638913976954-8f7b612867c2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2670&q=80'
    },
    text: {
      acceptSelectedText: 'custom text',
      acceptText: 'custom text',
      bannerLinkLabel: 'custom text',
      descriptionText: 'custom text',
      panelTitle: 'custom text',
      preferencesText: 'custom text',
      rejectText: 'custom text',
      saveButton: 'custom text',
      saveAllButton: 'custom text',
      servicesTag: 'custom text'
    },
    style: {
      accept: 'mt-2 text-green-600',
      bannerText: 'text-green-800',
      bannerTitle: '"text-white',
      closeButton: 'text-sky-100 bg-black',
      lockIcon: "text-red-500",
      panelHeader: 'md:flex justify-between px-4 py-4',
      panelText: 'text-red-300',
      panelTitle: 'text-white',
      preferencesText: 'text-black mt-4 border-0',
      privacyLink: 'text-black font-bold text-sm',
      reject: 'text-red-800 font-semibold text-xs px-2 py-0.5 mt-2',
      saveButton: 'text-md font-bold bg-white text-red-800 hover:bg-red-900',
      saveAllButton: 'text-md font-bold bg-white text-red-800 hover:bg-red-900',
      servicesTag: 'text-white',
      servicesText: 'text-yellow-400',
      stripes:'odd:bg-green-100 even:bg-green-800',
      toggles: 'bg-red-800',
    },
    activate: {
      GoogleTagManager: {
        container_id: 'GTM-EXAMPLE',
        dataJgcTag:"necessary", // The JGC tag of this script
        dataJgcService:"GoogleTagManager", // Custom name
        event_name: 'jgc-cookies-accepted', // example
        variables: [
            ['author', 'test'], // example
            ['price', 'test'] // example
        ]
      },
      GoogleAnalytics: {
          id: 'G-EXAMPLE', // example
          anonymized: true,
          dataJgcTag:"necessary", // cookie tag
          dataJgcService:"Google Analytics", // custom name
          ad_storage: false,
          analytics_storage: false
      },
      FacebookPixel: {
          init: 'EXAMPLE', // example
          dataJgcTag: 'functionalities', // cookie tag
          dataJgcService:"Facebook Pixel", // custom name
      },
    },
    panel: {
      bgColor: 'bg-red-900',
      open: true,
      padding: false,
    },
  });
</script>
```

<span name="main"></span>

## Main Configuration

Here you find the basic (and most important) parameters

Parameters:

| Parameter                                            | Type    | Default         | Values                                                                                   | Scope                                                                                                                             | Mandatory |

CHANGELOG.md
# Release Notes

## 0.9.9 (2022-06-12)

### What's improved

- Fixed a bug that prevented the banner from closing on certain mobile browsers

## 0.9.8 (2022-06-26)

### What's improved

- Minor fix

## 0.9.7 (2022-05-16)

### What's improved

- Bug Fixing

## 0.9.5 (2022-05-12)

### What's improved

- Bug Fixing

## 0.9.4 (2022-05-02)

### What's improved

- Bug Fixing

## 0.9.3.5 (2022-04-29)

### What's improved

- Bug Fixing

## 0.9.3.3 (2022-04-20)

### What's improved

- Bug Fixing

## 0.9.2 (2022-04-14)

### What's improved

- Refactoring (code-splitting)

<hr/>

## 0.9.1 (2022-04-02)

### What's new

- Added new locales: 🇫🇷 fr, 🇪🇸 es, 🇩🇪 de (but hey, I need your help to improve them)
- Added a new parameter for the style object: `preferenceDiv` (as you can imagine now you can customize the preference panel div).

### What's improved

- Fixed some minor bugs and added some new CSS classes

<hr>

## 0.9.0 (2022-04-01)

### What's new

- This changelog 😄

### What's improved

- [Now you can translate all the custom strings](https://github.com/francescomugnai/just-good-cookies/wiki/How-to-translate-the-custom-strings-and-make-JGC-fully-multilingual) <br/>
- Removed a useless function and changed some "checks"


src/justgoodcookies.js
/*
    @license
    JustGoodCookies
    Created by Francesco Mugnai 
    2022 - v 0.9.9
    Released under MIT License
    If you use this script, you will always remain the sole responsible party, use it at your own risk
    https://github.com/francescomugnai/just-good-cookies
*/


import { googleTagManager, checkActivations }  from "./activate";
import { autoMode, generatePreferenceStorage, checkCookiesAutoMode }  from "./autoMode";
import { showBanner, closeBanner }  from "./banner";
import { getCookie, getCookieId, getCookiePreferences, saveCookie, checkCookieExpiration, saveCookiesPreferences }  from "./cookies";
import { locales }  from "./locales";
import { activateToggledCookies, hideScripts, removeScript, removeDivsOfUserAcceptedIframes, checkGoogleAnalytics }  from "./scripts";
import { loadBannerLayout }  from "./styles";
import { removePlaceholders }  from "./placeholders";
import { closePreferencePanel, closePreferencePanelAndSaveAll, setPreferences, loadPreferences, openPanel  }  from "./preferences";
import { isString, isBoolean, isFunction, checkTailwindPrefix, checkDarkMode, checkBackground, loadText  }  from "./utilities";

class JustGoodCookies {
  constructor() {
    this.activate = undefined // Custom Activations
    this.auto = false // autoMode
    this.autoCategories = undefined // Categories for autoMode
    this.banner = undefined // Banner div
    this.bannerConfig = undefined // Banner config
    this.bannerLink = undefined // Privacy policy link
    this.bannerText = undefined // Custom banner text
    this.cookieTimeout = undefined; // Default cookie duration (360 days)
    this.config = undefined // General config
    this.customStyle = undefined // Banner style 
    this.darkMode = undefined // Force dark mode
    this.getCustomCookies = undefined  // Cookie preferences and tags
    this.locale = undefined // Locale 
    this.localeString = undefined // Lang string
    this.onAccept = undefined // Callback on "accept"
    this.onReject = undefined // Callback on "reject"
    this.panel = undefined // Preference panel
    this.positions = {} // Banner positions
    this.panelFooter = undefined // "Footer"
    this.panelHeader = undefined // "Header"
    this.placeholder = undefined // Placeholder
    this.tailwindPrefix = '' // Tailwind Prefix
    this.text = undefined // Custom texts

    this.getCookieId = getCookieId
    this.getCookiePreferences = getCookiePreferences

    String.prototype.escape = function() {
      var replace = {
        '>': '&gt;',
        '<': '&lt;',
        '&': '&amp;'
      };
      return this.replace(/[&<>]/g, (tag) => replace[tag] || tag);
    };
  }

  /*
  *  Check the current status
  */
  checkCookies() {
    let preference = getCookie('JgcPreferences');
    if(preference['duration']){
      let getValue = preference['duration'].value
      switch (getValue) {
        case "0":  // Cookies rejected :(
          if(this.auto) autoMode() 
          checkCookieExpiration() // Check if cookie is expired
          hideScripts() // Hide the scripts
          break;
        case "1": // Cookies accepted :)
          bannerContent.classList.add(checkTailwindPrefix('hidden')); 
          checkCookiesAutoMode() // Check if we are running the autoMode
          removePlaceholders() // Remove placeholders
          removeDivsOfUserAcceptedIframes() // Remove hidden divs (if any) for accepted cookies
          checkCookieExpiration()  // Check if the cookie is expired
          checkActivations() // Check if we need to activate some pre-built scripts
          checkGoogleAnalytics() // Check Google Analytics
          activateToggledCookies() // We enable cookies and manage them through the settings panel
          googleTagManager()  // Check if we need to turn on Google Tag Manager
          closeBanner() // Close the banner
          break;
      }
    } else { 
      // The banner has not been accepted yet, let's turn off all scripts and show the banner
      if(this.auto) autoMode() 
      hideScripts() 
      showBanner() 
    }
  }

  /*
  * Accepts cookies
  */
  yesCookies() {
    checkCookieExpiration('1') 
    checkActivations() 
    setPreferences() 
    activateToggledCookies() 
    closePreferencePanelAndSaveAll()
    let checkPreferences = getCookie('JgcPreferences');
    if(checkPreferences['remove'] > 0){
      removeScript(true) // We need to remove them AND refresh the page
    } else {
      googleTagManager() // Do not trigger Google Tag Manager twice
      removePlaceholders() 
    }
    if(this.onAccept && typeof(this.onAccept) == 'function') this.onAccept()
    closeBanner() 
    removeDivsOfUserAcceptedIframes()
    if(!this.bannerConfig.onAccept) window.location.reload()
    if(document.getElementById('preferenceDiv')) closePreferencePanel()
  }
  
  /*
  * Reject cookies
  */
  noCookies() {
    checkCookieExpiration('0') 
    let getPreferences = getCookie('JgcPreferences');
    const saveObj = {...getPreferences}
    saveCookie(saveObj)
    if(this.onReject && typeof(this.onReject) == 'function') this.onReject()
    closeBanner() 
  }

  /**
  * Activate the JGC engine and all the main functions
  */
  init(data) {
    data ? data : {}

    // Initialize the language
    if(data.locale){
      this.locales = locales
      this.locale = this.locales[data.locale.escape()] || this.locales['en']
      this.localeString = data.locale
    } 

    // Check if the autoMode is active or not
    if(data.autoMode && isBoolean(data.autoMode, "autoMode")){
      let checkPreferences = getCookie('JgcPreferences');
      if(!checkPreferences){
        let scripts = document.querySelectorAll('iframe,script')
        for(const element of scripts) element.classList.add(checkTailwindPrefix('hidden'))
      }
      this.auto = true 
      if(data.autoCategories){
        for(let a of Object.keys(data.autoCategories)){
          data.autoCategories[a][0] = data.autoCategories[a][0].escape()
          data.autoCategories[a][1] = data.autoCategories[a][1].escape()
        }
        this.autoCategories = data.autoCategories
      }
    }

    // General config
    this.config = {
      locale: data.locale !== undefined ? data.locale.escape() : new Languages('en'),
      layout: data.layout.escape() || 'style1',
      privacyLink: data.privacyLink.escape() || '',
    }

    // Cookie duration
    this.cookieTimeout = data.cookieDuration ? data.cookieDuration : 360

    // Tailwind Prefix
    if(data.tailwindPrefix){
      this.tailwindPrefix = data.tailwindPrefix
    } 

    // Automatic Dark Mode 
    if(data.dark){
      if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
      const htmlClass = document.querySelector('html')
      htmlClass.classList.add(checkTailwindPrefix('dark'))
      this.darkMode = true
     }
    }

    // Custom texts 
    this.text = {
      acceptSelectedText: data.text?.acceptSelectedText ? isString(data.text.acceptSelectedText, "acceptSelectedText") : this.locale.acceptSelectedText,
      acceptText: data.text?.acceptText ? isString(data.text.acceptText, "acceptText") : this.locale.acceptText,
      bannerLinkLabel: data.text?.bannerLinkLabel ? isString(data.text.bannerLinkLabel, "bannerLinkLabel") : this.locale.bannerLinkLabel,
      descriptionText: data.text?.descriptionText ? isString(data.text.descriptionText, "descriptionText") : null,
      panelTitle: data.text?.panelTitle ? isString(data.text.panelTitle, "panelTitle") : this.locale.panelTitle,
      preferencesText: data.text?.preferencesText ? isString(data.text.preferencesText, "preferencesText") : this.locale.preferencesText,
      rejectText: data.text?.rejectText ? isString(data.text.rejectText, "rejectText") : this.locale.rejectText,
      saveButton: data.text?.saveButton ? isString(data.text.saveButton, "saveButton") : this.locale.saveAndContinue,
      saveAllButton: data.text?.saveAllButton ? isString(data.text.saveAllButton, "saveAllButton") : this.locale.saveAndContinueAcceptAll,
      servicesTag: data.text?.servicesTag ? isString(data.text.servicesTag, "servicesTag") : this.locale.servicesText,
    }
  
    // Banner config & style 
    this.bannerConfig = {
      animation: isBoolean(data.banner?.animation, "animation") ? data.banner.animation : true,
      backgroundColor: data.banner?.backgroundColor ? isString(data.banner.backgroundColor, "backgroundColor") : checkTailwindPrefix('bg-white dark:bg-gray-800'),
      backgroundDark: data.banner?.backgroundDark ?isBoolean( data.banner.backgroundDark, "backgroundDark") : false,
      backgroundImage: data.banner?.backgroundImage ? isString(data.banner.backgroundImage, "backgroundImage") : null,
      closeButton: data.banner?.closeButton ? isBoolean(data.banner.closeButton, "closeButton") : true,
      closeButtonAccept: data.banner?.closeButtonAccept ? isBoolean(data.banner.closeButtonAccept, "closeButtonAccept") : false,
      disableReject: data.banner?.disableReject ? isBoolean(data.banner.disableReject, "disableReject") : false,
      icon: data.banner?.icon ? isString(data.banner.icon, "icon") : null,
      iconDark: data.banner?.iconDark ? isString(data.banner.iconDark, "iconDark") : null,
      innerBackgroundImage: data.banner?.innerBackgroundImage ? isString(data.banner.innerBackgroundImage, "innerBackgroundImage") : null,
      logo: data.banner?.logo ? isString(data.banner.logo, "logo") : undefined,
      logoClasses: data.banner?.logoClasses ? isString(data.banner.logoClasses, "logoClasses") : undefined,
      maxWidth: data.banner?.maxWidth ? isString(data.banner.maxWidth, "maxWidth") : undefined,
      onAccept: data.banner?.onAccept ? this.onAccept = isFunction(data.banner.onAccept, "onAccept") : null,
      onReject: data.banner?.onReject ? this.onReject = isFunction(data.banner.onReject, "onReject") : null,
      position: data.banner?.position ? isString(data.banner.position) : undefined,
      shortText: data.banner?.shortText && isBoolean(data.banner.shortText, "shortText") ? this.locale.acceptShortText : this.acceptText,
      title: data.banner?.title ? isString(data.banner.title, "title") : 'Cookies',
    }

    // Custom text placeholder
    if(data.placeholder) {
      this.placeholder = {
        classes: data.placeholder?.classes && isString(data.placeholder.classes, "placeholder classes"),
        image: data.placeholder?.image && isString(data.placeholder.image, "placeholder image"),
        text: data.placeholder?.text && isString(data.placeholder.text, "placeholder text"),
      }
    }
   
    // Preference Panel
    if(data.panel) {
      this.panel = {
        bgColor: data.panel?.bgColor ? isString(data.panel.bgColor, "bgColor") : null,
        open: data.panel?.open ? isBoolean(data.panel.open, "open") : false,
        padding: data.panel?.padding ? isBoolean(data.panel.padding, "padding") : false,
      }
    }

    // Banner style
    if(data.style) {
      this.customStyle = {        
        accept: data.style?.accept ? isString(data.style.accept, "accept") : null,
        bannerText: data.style?.bannerText ? isString(data.style.bannerText, "bannerText") : null,
        bannerTitle: data.style?.bannerTitle ? isString(data.style.bannerTitle, "bannerTitle") : null,
        closeButton: data.style?.closeButton ? isString(data.style.closeButton, "closeButton") : null, 
        toggles: data.style?.toggles ? isString(data.style.toggles, "toggles") : null,
        lockIcon: data.style?.lockIcon ? isString(data.style.lockIcon, "lockIcon") : null,
        panelHeader: data.style?.panelHeader ? isString(data.style.panelHeader, "panelHeader") : null, 
        panelText: data.style?.panelText ? isString(data.style.panelText, "panelText") : null, 
        panelTitle:  data.style?.panelTitle ? isString(data.style.panelTitle, "panelTitle") : null,
        preferenceDiv:  data.style?.preferenceDiv ? isString(data.style.preferenceDiv, "preferenceDiv") : null,
        preferencesText: data.style?.preferencesText ? isString(data.style.preferencesText, "preferencesText") : null, 
        privacyLink: data.style?.privacyLink ? isString(data.style.privacyLink, "privacyLink") : null, 
        reject: data.style?.reject ? isString(data.style.reject, "reject") : null,
        saveButton: data.style?.saveButton ? isString(data.style.saveButton, "classes") : null,
        saveAllButton: data.style?.saveAllButton ? isString(data.style.saveAllButton, "classes") : null,
        servicesText: data.style?.servicesText ? isString(data.style.servicesText, "servicesText") : null, 
        servicesTag: data.style?.servicesTag ? isString(data.style.servicesTag, "servicesTag") : null, 
        stripes: data.style?.stripes ? isString(data.style.stripes, "stripes: classes") : null,
      }
    }
  
    // Cookie Categories
    if(data.cookies) {
      this.getCustomCookies = {}
      for(let jgcTag of Object.keys(data.cookies)){
        for (let [objKey] of Object.entries(data.cookies[jgcTag])) data.cookies[jgcTag][objKey] = isString(data.cookies[jgcTag][objKey])
      }
      this.getCustomCookies = data.cookies
    }

    // Activations
    this.activate = data.activate ? data.activate : null
    
    // Let's start the engine
    if(document.readyState == 'complete' || document.readyState == 'loading') {
      checkDarkMode() // Check Dark Mode
      loadText() // Check if there is a custom text for the banner
      generatePreferenceStorage() // Create the default user settings
      checkBackground() // Check if we need to add a dark overlay
      loadPreferences() // Make the preference button clickable
      saveCookiesPreferences() // Save cookies
      loadBannerLayout(this.config.layout) // Load the banner
      openPanel() // Check whether the preferences panel should be visible or not
      this.checkCookies() // Check cookies
    }
  }
}


export default new JustGoodCookies()